### PR TITLE
PersistedAssemblyBuilder missing methods on generated types

### PR DESCRIPTION
**Pull request by snakex64**
#108733 

Add missing methods to PersistedAssemblyBuilder generated types:

- GetProperties
- GetProperty
- GetInterface

Edit:
- GetNestedType
- GetNestedTypes
- GetMember
- GetMembers

Also provide bugfixes to GetMethods, GetFields, GetProperties, GetInterfaces not returning parent members

### CODE REVIEW

**Comment by snakex64** [comment]
@dotnet-policy-service agree

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs [hunk 1]
```diff
@@ -891,18 +891,130 @@ public override FieldInfo[] GetFields(BindingFlags bindingAttr)
...
+                {
+                    candidates.Add(propertyInfo);
+                }
+            }
+
+            return candidates;
+        }
 
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicNestedTypes | DynamicallyAccessedMemberTypes.NonPublicNestedTypes)]
         public override Type[] GetNestedTypes(BindingFlags bindingAttr) => throw new NotSupportedException();
```
[the remark this answers is not in the document]
Comment (snakex64) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs:1020]: Seems like it is available by going through the module. Should be easy. Will try to add that one too tomorrow. I'm not sure of all the different options of mapping between BindingFlags and the attribute enum used to create types so I will create a base version and we can augment it after

[the remark this answers is not in the document]
Comment (snakex64) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs]: Alright, should be good now if you want to have a look

**Comment by snakex64** [comment]
Added GetMember and GetMembers too. It made me realize that there was some bugs related to parent members not being returned by GetMethods, GetFields, etc... By default unless `BindingFlags.DeclaredOnly` is specified, the values of the parent class should also be returned. I added a potential bugfix to all of them

**Comment by snakex64** [comment]
Considering this is not just a new feature anymore but more of a bugfix to existing APIs such as `GetFields`, do ou think it would be possible to target the .NET 9 release so we can at least fix those last second ?

File: src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs [hunk 2]
```diff
@@ -660,7 +660,7 @@ public void DefineMethodOverride(System.Reflection.MethodInfo methodInfoBody, Sy
         [System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.NonPublicProperties | System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicProperties)]
         public override System.Reflection.PropertyInfo[] GetProperties(System.Reflection.BindingFlags bindingAttr) { throw null; }
         [System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.NonPublicProperties | System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicProperties)]
-        protected override System.Reflection.PropertyInfo GetPropertyImpl(string name, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder? binder, System.Type? returnType, System.Type[]? types, System.Reflection.ParameterModifier[]? modifiers) { throw null; }
+        protected override System.Reflection.PropertyInfo? GetPropertyImpl(string name, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder? binder, System.Type? returnType, System.Type[]? types, System.Reflection.ParameterModifier[]? modifiers) { throw null; }
```
Comment (steveharter) [inline src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs:663]: Not sure why we don't get compile warning for this; the base `Type.GetPropertyImpl` returns PropertyInfo?.

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs [hunk 3]
```diff
@@ -775,12 +793,30 @@ public override ConstructorInfo[] GetConstructors(BindingFlags bindingAttr)
...
+            return methods.ToArray();
+        }
+
         [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)]
         public override MethodInfo[] GetMethods(BindingFlags bindingAttr)
         {
             ThrowIfNotCreated();
 
-            List<MethodInfo> methods = new();
+            List<MethodInfo> methods = new List<MethodInfo>();
```
Comment (steveharter) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs:819]: nit: unnecessary change here; using `new()` is fine.

**Review by steveharter — APPROVED** [review]
LGTM; thanks for the contribution.
